Code/GraphMol/Descriptors/MORSE.cpp

```diff
@@ -0,0 +1,365 @@
+  int confId = conf.getId();
+
+  std::vector<double> R = getG(32);
+  std::vector<double> RDFres;
+  double *DM = MolOps::get3DDistanceMat(mol, confId);
+
+  for (int i = 0; i < 32; i++) {
+    double res = 0;
+    for (int j = 0; j < numAtoms - 1; j++) {
+      for (int k = j + 1; k < numAtoms; k++) {
+
+        if (i==0) { res += 1;}
+
+        else
+
+          res += sin(R[i] * DM[j * numAtoms + k]) / (R[i] * DM[j * numAtoms + k]);
+      }
+    }
+
+    RDFres.push_back(round( 1000 * res) / 1000);
+  }
+
+  return RDFres;
+}
+
+
+std::vector<double> CalcMassMORSE(const ROMol &mol, const Conformer &conf) {
+  int numAtoms = conf.getNumAtoms();
+  int confId = conf.getId();
+
+  std::vector<double> R = getG(32);
+  std::vector<double> RDFres;
+  double *DM = MolOps::get3DDistanceMat(mol, confId);
+
+  
+  std::vector<double> mass = moldata3D.GetRelativeMW(mol);
+
+  for (int i = 0; i < 32; i++) {
+    double res = 0;
+    for (int j = 0; j < numAtoms - 1; j++) {
+      for (int k = j + 1; k < numAtoms; k++) {
+
+       
+        if (i==0) { res += mass[j] * mass[k];}
+
+        else
+
+        res += mass[j] * mass[k] * sin(R[i] * DM[j * numAtoms + k]) / (R[i] * DM[j * numAtoms + k]);
+      }
+    }
+
+    RDFres.push_back(round(1000 * res ) / 1000);
+  }
+
+
+
+  return RDFres;
+}
+
+std::vector<double> CalcAtomNumMORSE(
+    const ROMol &mol, const Conformer &conf) {
+  int numAtoms = conf.getNumAtoms();
+  int confId = conf.getId();
+
+  std::vector<double> R = getG(32);
+  std::vector<double> RDFres;
+
+  double *DM = MolOps::get3DDistanceMat(mol, confId);
+
+  std::vector<int> AN;
+  for (int p = 0; p < numAtoms; p++) {
+    AN.push_back(mol.getAtomWithIdx(p)->getAtomicNum());
+  }
+
+  for (int i = 0; i < 32; i++) {
+    double res = 0;
+    for (int j = 0; j < numAtoms - 1; j++) {
+      for (int k = j + 1; k < numAtoms; k++) {
+       
+        if (i==0) { res += AN[j] * AN[k];}
+
+        else
+
+        res += AN[j] * AN[k] * sin(R[i] * DM[j * numAtoms + k]) / (R[i] * DM[j * numAtoms + k]);
+      }
+    }
+    // 36 = 6*6 Atomic number of the Carbon
+    
+    RDFres.push_back(round( 1000 * res / 36) / 1000);
+  }
+
+  return RDFres;
+}
+
+std::vector<double> CalcIonPolMORSE(const ROMol &mol, const Conformer &conf) {
+  int numAtoms = conf.getNumAtoms();
+  int confId = conf.getId();
+
+  std::vector<double> R = getG(32);
+  std::vector<double> RDFres;
+  double *DM = MolOps::get3DDistanceMat(mol, confId);
+
+  std::vector<double> RelativeIonPol = moldata3D.GetRelativeIonPol(mol);
+
+  for (int i = 0; i < 32; i++) {
+    double res = 0;
+    for (int j = 0; j < numAtoms - 1; j++) {
+      for (int k = j + 1; k < numAtoms; k++) {
+     
+        if (i==0) { res += RelativeIonPol[j] * RelativeIonPol[k];}
+
+        else
+        res += RelativeIonPol[j] * RelativeIonPol[k] * sin(R[i] * DM[j * numAtoms + k]) /
+               (R[i] * DM[j * numAtoms + k]);
+      }
+    }
+
+    RDFres.push_back(round( 1000 * res) / 1000);
+  }
```

Comment:
This is all very repetitive code. With the exception of `CalcUnweightedMorse`, these functions have a highly similar structure. What about a `CalcWeightedMorse` function.
```
std::vector<double> CalcWeightedMORSE(const ROMol &mol, const Conformer &conf, const std::vector<double> weight) {
    int numAtoms = conf.getNumAtoms();
    int confId = conf.getId();

   std::vector<double> R = getG(32);
   std::vector<double> RDFres;
   double *DM = MolOps::get3DDistanceMat(mol, confId);

   for (int i = 0; i < 32; i++) {
     double res = 0;
     for (int j = 0; j < numAtoms - 1; j++) {
       for (int k = j + 1; k < numAtoms; k++) {
         if (i==0) { 
           res += weight[j] * weight[k];
         } else {
           res += weight[j] * weight[k] * sin(R[i] * DM[j * numAtoms + k]) / (R[i] * DM[j * numAtoms + k]);
         }
       }
     }
     RDFres.push_back(round( 1000 * res) / 1000);
  }
  return RDFres;
}
```
The specific Morse functions can than make use of this:
```
std::vector<double> CalcIonPolMORSE(const ROMol &mol, const Conformer &conf) {
   std::vector<double> RelativeIonPol = moldata3D.GetRelativeIonPol(mol);
   return CalcWeightedMORSE(mol, conf, weight);
}

std::vector<double> CalcPolMORSE(const ROMol &mol, const Conformer &conf) {
   std::vector<double> weight = moldata3D. GetRelativePol(mol);
   return CalcWeightedMORSE(mol, conf, weight);
}
```